File: packages/terminal-native/android/src/main/java/dev/happier/terminal/HappierTerminalNativeModule.kt
package dev.happier.terminal

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Queues

class HappierTerminalNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HappierTerminalNative")

    Events("rendererCrash", "surfaceReady", "writeAck", "input", "resize", "link", "selection", "copy", "title", "bell")

    Function("getAvailability") {
      return@Function TermuxBridge.availability()
    }

    Function("getQaCapabilities") {
      return@Function mapOf(
        "rendererCrashInjection" to BuildConfig.HAPPIER_TERMINAL_NATIVE_QA_CRASH_INJECTION,
      )
    }

    Function("getAndroidTermuxDiagnostics") {
      val diagnostic = TermuxBridge.diagnostic()
      return@Function mapOf(
        "available" to diagnostic.available,
        "renderer" to diagnostic.renderer,
        "reason" to diagnostic.reason,
        "detail" to diagnostic.detail,
        "fallbackRenderer" to diagnostic.fallbackRenderer,
        "fallbackRequired" to diagnostic.fallbackRequired,
        "engineeringQaOverride" to diagnostic.engineeringQaOverride,
        "requiredModules" to diagnostic.requiredModules,
        "forbiddenModules" to diagnostic.forbiddenModules,
        "remoteSessionAdapterRequired" to diagnostic.remoteSessionAdapterRequired,
        "blockers" to diagnostic.blockers.map { mapOf("reason" to it.reason, "detail" to it.detail) },
      )
    }

    AsyncFunction("createSurface") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.createSurface(surfaceId) { eventName, payload ->
        sendEvent(eventName, payload)
      }
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("qaInjectRendererCrash") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.qaInjectRendererCrash(surfaceId)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("writeBytes") { surfaceId: String, base64Bytes: String, byteOffset: Double ->
      return@AsyncFunction TermuxBridge.writeBytes(surfaceId, base64Bytes, byteOffset.toLong())
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("sendInputBytes") { surfaceId: String, base64Bytes: String ->
      return@AsyncFunction TermuxBridge.sendInputBytes(surfaceId, base64Bytes)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("resizeSurface") { surfaceId: String, cols: Int, rows: Int ->
      return@AsyncFunction TermuxBridge.resizeSurface(surfaceId, cols, rows)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("focusSurface") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.focusSurface(surfaceId)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("clearSurface") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.clearSurface(surfaceId)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("disposeSurface") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.disposeSurface(surfaceId)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("copySelection") { surfaceId: String ->
      return@AsyncFunction TermuxBridge.copySelection(surfaceId)
    }.runOnQueue(Queues.MAIN)

    View(TermuxView::class) {
      Name("HappierTerminalNativeView")

      Prop("surfaceId") { view: TermuxView, surfaceId: String ->
        view.setSurfaceId(surfaceId)
      }

      Prop("fontSize") { view: TermuxView, fontSize: Double ->
        view.setTerminalFontSize(fontSize)
      }

      Prop("lineHeightPx") { view: TermuxView, lineHeightPx: Double ->
        view.setTerminalLineHeightPx(lineHeightPx)
      }

      Prop("accessibilitySummary") { view: TermuxView, accessibilitySummary: String ->
        view.setAccessibilitySummary(accessibilitySummary)
      }

      Prop("accessibilityAccepted") { view: TermuxView, accessibilityAccepted: Boolean ->
        view.setAccessibilityAccepted(accessibilityAccepted)
      }

      Prop("accessibilityTerminalLabel") { view: TermuxView, label: String ->
        view.setAccessibilityTerminalLabel(label)
      }

      Prop("accessibilityFallbackValue") { view: TermuxView, value: String ->
        view.setAccessibilityFallbackValue(value)
      }

      Prop("accessibilityFocusActionLabel") { view: TermuxView, label: String ->
        view.setAccessibilityFocusActionLabel(label)
      }

      Prop("accessibilityCopySelectionActionLabel") { view: TermuxView, label: String ->
        view.setAccessibilityCopySelectionActionLabel(label)
      }

      Prop("accessibilitySelectAllActionLabel") { view: TermuxView, label: String ->
        view.setAccessibilitySelectAllActionLabel(label)
      }

      Prop("accessibilityOpenLinkActionLabel") { view: TermuxView, label: String ->
        view.setAccessibilityOpenLinkActionLabel(label)
      }
    }

    OnDestroy {
      TermuxBridge.disposeAllOnMain()
    }
  }
}

File: packages/terminal-native/ios/HappierTerminalNativeModule.swift
import ExpoModulesCore
import Foundation

public final class HappierTerminalNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HappierTerminalNative")

    Events("rendererCrash", "surfaceReady", "writeAck", "input", "resize", "link", "selection", "copy", "title", "bell")

    Function("getAvailability") { () -> [String: Any] in
      makeGhosttyRuntimeDiagnostic().availabilityPayload()
    }

    AsyncFunction("createSurface") { (surfaceId: String) async -> [String: Any] in
      await MainActor.run {
        let diagnostic = makeGhosttyRuntimeDiagnostic()
        guard diagnostic.isAvailable else {
          return diagnostic.availabilityPayload()
        }
        guard let view = GhosttySurfaceRegistry.shared.surface(id: surfaceId),
              view.prepareSurface() else {
          return [
            "available": false,
            "reason": "surface-not-ready",
            "detail": "Native terminal surface is not mounted or has no drawable size.",
          ]
        }
        return diagnostic.availabilityPayload()
      }
    }

    AsyncFunction("writeBytes") { (surfaceId: String, base64Bytes: String, byteOffset: Int64) async -> [String: Any] in
      await MainActor.run {
        guard let view = GhosttySurfaceRegistry.shared.surface(id: surfaceId) else {
          return [
            "accepted": false,
            "reason": "surface-not-ready",
            "detail": "Native terminal surface is not mounted.",
          ]
        }
        guard let bytes = Data(base64Encoded: base64Bytes) else {
          return [
            "accepted": false,
            "reason": "invalid-ack",
            "detail": "Native terminal write payload was not valid base64.",
          ]
        }
        return view.writeBytes(bytes, byteOffset: byteOffset)
      }
    }

    AsyncFunction("resizeSurface") { (surfaceId: String, cols: Int, rows: Int) async -> Void in
      await MainActor.run {
        GhosttySurfaceRegistry.shared.surface(id: surfaceId)?.resize(cols: cols, rows: rows)
      }
    }

    AsyncFunction("focusSurface") { (surfaceId: String) async -> Void in
      await MainActor.run {
        GhosttySurfaceRegistry.shared.surface(id: surfaceId)?.focusSurface()
      }
    }

    AsyncFunction("clearSurface") { (surfaceId: String) async -> Void in
      await MainActor.run {
        GhosttySurfaceRegistry.shared.surface(id: surfaceId)?.clearSurface()
      }
    }

    AsyncFunction("disposeSurface") { (surfaceId: String) async -> Void in
      await MainActor.run {
        GhosttySurfaceRegistry.shared.surface(id: surfaceId)?.disposeSurface()
      }
    }

    AsyncFunction("copySelection") { (surfaceId: String) async -> [String: Any] in
      await MainActor.run {
        GhosttySurfaceRegistry.shared.surface(id: surfaceId)?.copySelection() ?? [
          "copied": false,
          "reason": "surface-not-ready",
        ]
      }
    }

    View(GhosttySurfaceView.self) {
      ViewName("HappierTerminalNativeView")

      Prop("surfaceId") { [weak self] (view: GhosttySurfaceView, surfaceId: String) in
        view.setEventEmitter { [weak self] eventName, payload in
          self?.sendEvent(eventName, payload)
        }
        view.surfaceId = surfaceId
      }

      Prop("fontSize", 14.0) { (view: GhosttySurfaceView, fontSize: Double) in
        view.fontSize = fontSize
      }

      Prop("lineHeightPx", 18.0) { (view: GhosttySurfaceView, lineHeightPx: Double) in
        view.lineHeightPx = lineHeightPx
      }

      Prop("accessibilitySummary") { (view: GhosttySurfaceView, accessibilitySummary: String) in
        view.accessibilitySummary = accessibilitySummary
      }

      Prop("accessibilityAccepted", false) { (view: GhosttySurfaceView, accessibilityAccepted: Bool) in
        view.accessibilityAccepted = accessibilityAccepted
      }

      Prop("accessibilityTerminalLabel") { (view: GhosttySurfaceView, value: String) in
        view.accessibilityTerminalLabel = value
      }

      Prop("accessibilityFallbackValue") { (view: GhosttySurfaceView, value: String) in
        view.accessibilityFallbackValue = value
      }

      Prop("accessibilityFocusActionLabel") { (view: GhosttySurfaceView, value: String) in
        view.accessibilityFocusActionLabel = value
      }

      Prop("accessibilityCopySelectionActionLabel") { (view: GhosttySurfaceView, value: String) in
        view.accessibilityCopySelectionActionLabel = value
      }
    }
  }
}
